src/Compilers/CSharp/Portable/Parser/LanguageParser_Patterns.cs

```diff
@@ -222,51 +223,242 @@ private bool ScanDesignation(bool permitTuple)
 
         // Priority is the ExpressionSyntax. It might return ExpressionSyntax which might be a constant pattern such as 'case 3:' 
         // All constant expressions are converted to the constant pattern in the switch binder if it is a match statement.
-        // It is used for parsing patterns in the switch cases. It never returns constant pattern!
-        private CSharpSyntaxNode ParseExpressionOrPatternForCase()
+        // It is used for parsing patterns in the switch cases. It never returns constant pattern, because for a `case` we
+        // need to use a pre-pattern-matching syntax node for a constant case.
+        private CSharpSyntaxNode ParseExpressionOrPattern(bool forCase)
         {
-            var tk = this.CurrentToken.Kind;
-            CSharpSyntaxNode node = null;
+            // handle common error recovery situations during typing
+            switch (this.CurrentToken.Kind)
+            {
+                case SyntaxKind.CommaToken:
+                case SyntaxKind.SemicolonToken:
+                case SyntaxKind.CloseBraceToken:
+                case SyntaxKind.CloseParenToken:
+                case SyntaxKind.CloseBracketToken:
+                case SyntaxKind.EqualsGreaterThanToken:
+                    return this.ParseIdentifierName(ErrorCode.ERR_MissingArgument);
+            }
 
-            // If it is a nameof, skip the 'if' and parse as an expression. 
-            if ((SyntaxFacts.IsPredefinedType(tk) || tk == SyntaxKind.IdentifierToken) &&
-                  this.CurrentToken.ContextualKind != SyntaxKind.NameOfKeyword)
+            var resetPoint = this.GetResetPoint();
+            try
             {
-                var resetPoint = this.GetResetPoint();
-                try
+                TypeSyntax type = null;
+                var tk = this.CurrentToken.Kind;
+                if ((SyntaxFacts.IsPredefinedType(tk) || tk == SyntaxKind.IdentifierToken) &&
```

Comment:
I'm not exactly clear on the structure here: when parsing a pattern like `C(int, int)` is `C` considered to be part of the deconstruction itself? Or is there a type pattern with a nested deconstruction?

Reply:
`C(int, int)` isn't a pattern, because `int` isn't a pattern.

Here is the grammar being parsed:

``` antlr
pattern
	: declaration_pattern
	| constant_pattern
	| deconstruction_pattern
	| property_pattern
	;
declaration_pattern
	: type identifier
	;
constant_pattern
	: expression
	;
deconstruction_pattern
	: type? '(' subpatterns? ')' property_subpattern? identifier?
	;
subpatterns
	: subpattern
	| subpattern ',' subpatterns
	;
subpattern
	: pattern
	| identifier ':' pattern
	;
property_subpattern
	: '{' subpatterns? '}'
	;
property_pattern
	: property_subpattern identifier?
	;
```

Reply:
I will add that to the internal documentation in the next PR.